providers/base/bin/gateway_ping_test.py

```diff
@@ -83,151 +108,297 @@ def _get_default_gateway_from_bin_route(self):
-                                            stderr=subprocess.STDOUT)
-                    break
-    except subprocess.CalledProcessError:
-        pass
-    ARP_POPULATE_TRIES = 10
-    num_tries = 0
-    while num_tries < ARP_POPULATE_TRIES:
+            return None
+        for line in network_info.splitlines():
+            line = line.strip()
+            if line.startswith("Gateway:"):
+                return line.split()[-1]
+        return None
+
+    def get_default_gateways(self) -> set:
+        """
+        Use multiple sources to get the default gateway to be robust to
+        possible platform bugs
+        """
+        def_gateways = {
+            self._get_default_gateway_from_ip(),
+            self._get_default_gateway_from_proc(),
+            self._get_default_gateway_from_bin_route(),
+            self._get_default_gateway_from_networkctl(),
+        }
+        def_gateways -= {None}
+        if len(def_gateways) > 1:
+            logging.warning(
+                "Found more than one default gateway for interface {}".format(
+                    self.interface
+                )
+            )
+        return def_gateways
+
+    @classmethod
+    def get_interface_from_ip(cls, ip):
+        # Note: this uses -o instead of -j for xenial/bionic compatibility
+        route_info = subprocess.check_output(
+            ["ip", "-o", "route", "get", ip], universal_newlines=True
+        )
+        for line in route_info.splitlines():
+            # ip dev device_name src ...
+            fields = line.split()
+            if len(fields) > 3:
+                return fields[2]
+        raise ValueError(
+            "Unable to determine any device used for {}".format(ip)
+        )
+
+    @classmethod
+    def get_any_interface(cls):
+        # Note: this uses -o instead of -j for xenial/bionic compatibility
+        route_infos = subprocess.check_output(
+            ["ip", "-o", "route", "show", "default", "0.0.0.0/0"],
+            universal_newlines=True,
+        )
+        for route_info in route_infos.splitlines():
+            route_info_fields = route_info.split()
+            if len(route_info_fields) > 5:
+                return route_info_fields[4]
+        raise ValueError("Unable to determine any valid interface")
+
+    @classmethod
+    def from_ip(cls, ip: str):
+        """
+        Build an instance of Route given an ip, if no ip is provided the best
+        interface that can route to 0.0.0.0/0 is selected (as described by
+        metric)
+        """
+        if ip:
+            interface = Route.get_interface_from_ip(ip)
+            return Route(interface)
+        return Route(Route.get_any_interface())
+
+
+def is_reachable(ip, interface, verbose=False):
+    """
+    Ping an ip to see if it is reachable
+    """
+    result = ping(ip, interface, 3, 10, verbose)
+    return result["transmitted"] >= result["received"] > 0
+
+
+def get_default_gateway_reachable_on(interface: str, verbose=False) -> str:
```

Comment:
The `verbose` seems not be used, do we still need it here?

Reply:
no, removed. You also made me spot an error in 3 lines up (verbose is a kwarg, should be passed as such, the order of params has changed)